### PR TITLE
Implement pinvoke with PreserveSig=false.

### PR DESCRIPTION
**Pull request by madewokherd**

### CODE REVIEW

File: mono/tests/marshalsig.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+		try {
+			mono_preserve_sig_test (unchecked((int)0x80004001) /* E_NOTIMPL */);
+			// should throw NotImplementedException
+			return 3;
+		}
+		catch (NotImplementedException e) { }
+
+		try {
+			mono_preserve_sig_test (unchecked((int)0x80004001) /* E_NOTIMPL */, out result);
```
Comment (zfigura) [inline mono/tests/marshalsig.cs:31]: Should the result get written on an exception? Should we test the value of 'result' here, or the return value in the other call?

(Also, in general initializing 'result' to some sentinel to make sure that it really is getting correctly written is probably not a bad idea).

File: mono/metadata/marshal.c [hunk 2]
```diff
@@ -3500,6 +3505,98 @@ mono_marshal_set_callconv_from_modopt (MonoMethod *method, MonoMethodSignature *
...
+	memcpy (sig_adjusted, sig, sigsize);
+	sig_adjusted->param_count = param_count;
+
+	if (has_result) {
+		sig_adjusted->params[sig->param_count] = mono_metadata_type_dup (image, sig->ret);
+		sig_adjusted->params[sig->param_count]->byref = 1;
+		sig_adjusted->params[sig->param_count]->attrs = PARAM_ATTRIBUTE_OUT;
+	}
+
+	sig_adjusted->ret = mono_get_int32_type();
```
Comment (zfigura) [inline mono/metadata/marshal.c:3538]: Shouldn't we be also setting sig_adjusted->call_convention somewhere in here?

**Review by zfigura — COMMENTED** [review]
Looks pretty straightforward to me. Most of this looks correct on account of looking like the cominterop code (and because that too looks correct). I hate the fact that this code is duplicated between cominterop and here, but if I started trying to refactor mono I'd be here a long time.

File: mono/tests/marshalsig.cs (hunk 1 above)
Comment (madewokherd) [reply mono/tests/marshalsig.cs:31]: Well, if the function throws an exception as it should, there is no return value, but the rest of that sounds like a good idea.

File: mono/metadata/marshal.c (hunk 2 above)
Comment (madewokherd) [reply mono/metadata/marshal.c:3538]: As I understand it, the memcpy should have copied that field.

**Comment by madewokherd** [comment]
> Looks pretty straightforward to me. Most of this looks correct on account of looking like the cominterop code (and because that too looks correct). I hate the fact that this code is duplicated between cominterop and here, but if I started trying to refactor mono I'd be here a long time.

Maybe the preservesig=false functionality could be moved out of cominterop wrappers in favor of this new function, but that's tricky because the cominterop code uses mono_marshal_emit_native_wrapper (with an embedded assumption that it's going to generate a managed-to-native wrapper). We'd either need a function to replace mono_marshal_emit_native_wrapper without having the caller pass in a MonoMethodBuilder (which they might also have a reason for doing), or we'd need to embed preservesig=false support into the managed-to-native wrapper, which was my initial approach but for some reason I gave up on that in favor of another wrapper..

File: mono/metadata/marshal.c (hunk 2 above)
Comment (zfigura) [reply mono/metadata/marshal.c:3538]: Oops, you're right, there is a memcpy there >_>

**Pull request merged**